### PR TITLE
Include nodejs and gomodule updaters

### PR DESCRIPTION
**Pull request by inode64**
Add 2 new modules to update the ebuilds

Along the way I have improved the logic of uploading to git

The digest is now generated automatically
I have changed so that updates can be multiple.
There are ebuilds that have both gomod and nodejs to process

I have added parameter control in the json file

It would be necessary to adjust the current updates so that they use the content of the main process, tell me if you see it correct and I will prepare it

### CODE REVIEW

File: livecheck/special/utils.py [hunk 1]
```diff
@@ -1,9 +1,98 @@
 from pathlib import Path
+from loguru import logger
```
Comment (Tatsh) [inline livecheck/special/utils.py:2]: I want to get away from loguru.

```suggestion
import logging
logger = logging.getLogger(__name__)
```

Put `logger = ...` after `__all__`.

File: livecheck/utils/portage.py [hunk 2]
```diff
@@ -231,3 +233,43 @@ def compare_versions(old: str, new: str) -> bool:
...
     return vercmp(sanitize_version(new), sanitize_version(old), silent=0) == -1
+
+
+def get_distdir() -> str:
+    settings = portage.config(clone=portage.settings)
+    distdir = settings.get('DISTDIR')
+    if distdir:
+        return distdir
+    else:
+        return '/var/cache/distfiles'
```
Comment (Tatsh) [inline livecheck/utils/portage.py:244]:
```suggestion
    return '/var/cache/distfiles'
```

File: livecheck/settings.py [hunk 3]
```diff
@@ -57,30 +61,44 @@ def gather_settings(search_dir: str) -> LivecheckSettings:
...
+                    ignored_packages.add(catpkg)
+                elif settings_parsed.get('type').lower() == 'regex':
+                    custom_livechecks[catpkg] = (settings_parsed['url'], settings_parsed['regex'],
+                                                 settings_parsed.get('use_vercmp', True),
+                                                 settings_parsed.get('version', None))
+                elif settings_parsed.get('type').lower() == 'checksum':
+                    checksum_livechecks.add(catpkg)
+                else:
+                    logger.error(
+                        f"Unknown \"type\" in {path}, only \"none\", \"regex\", and \"checksum\" are supported."
```
Comment (Tatsh) [inline livecheck/settings.py:89]:
```suggestion
                        f'Unknown "type" in {path}, only "none", "regex", and "checksum" are supported.'
```

File: livecheck/main.py [hunk 4]
```diff
@@ -38,6 +38,8 @@
 from .special.rubygems import get_latest_rubygems_package
 from .special.sourceforge import get_latest_sourceforge_package
 from .special.jetbrains import get_latest_jetbrains_package, update_jetbrains_ebuild
+from .special.gomodule import update_gomodule_ebuild, remove_gomodule_url
+from .special.nodejs import update_nodejs_ebuild, remove_nodejs_url
```
Comment (Tatsh) [inline livecheck/main.py:42]: Please sort the imports.

File: livecheck/main.py [hunk 5]
```diff
@@ -386,20 +381,48 @@ def do_main(*, auto_update: bool, keep_old: bool, cat: str, ebuild_version: str,
                     sp.run(('git', 'mv', ebuild, new_filename), check=True)
                 else:
                     sp.run(('mv', ebuild, new_filename), check=True)
-            else:
+            # Write net ebuild content because it may be modified
```
Comment (Tatsh) [inline livecheck/main.py:384]: Did you mean 'Write new ebuild content'?

**Comment by Tatsh** [comment]
Regarding distribution archives for Go and Node.js, where are you putting these? It almost seems like livecheck should handle the uploading of these to somewhere. GitHub releases API would be an example place. But it should be sort of generic.

**Review by Tatsh — CHANGES_REQUESTED** [review]

**Comment by inode64** [comment]
> Regarding distribution archives for Go and Node.js, where are you putting these? It almost seems like livecheck should handle the uploading of these to somewhere. GitHub releases API would be an example place. But it should be sort of generic.
> 
> The reason I ask is because I currently have my own way of dealing with Node.js packages without having to make tarballs and for Go packages I still use the old EGO_SUM method. It is easier than trying to maintain the tarballs at the moment.

Yes, I will prepare that in a new pull request, because now I am doing it manually in my overlay
Now, if the compressed file is small, I left it in git
https://github.com/inode64/inode64-overlay/tree/main/dist
, and if not, I upload it to my web server

That's why I had thought about creating a hook to be able to perform these tasks automatically.

It was to avoid adding too much in the same pull request

**Comment by inode64** [comment]
Do you see this code correctly? Do I have to modify something else?

**Comment by Tatsh** [comment]
There is a bug in this branch and in `master`. Ebuilds that have SHAs to update but not a different date according to Git have their ebuilds appear in the root of the project:

```
~/overlay $ ls -la *.ebuild
-rw-r--r-- 1 root root 682 2024-11-20 03:58:23 pymatting-1.1.12_p20241110-r1.ebuild
-rw-r--r-- 1 root root 468 2024-11-20 03:58:13 spirv-cross-20241113-r1.ebuild
-rw-r--r-- 1 root root 608 2024-11-20 03:58:29 vapoursynth-fsrcnn-ncnn-vulkan-20241112-r1.ebuild
```

The deletion of the old ebuild works but where the ebuild outputs is definitely wrong.

They should be in the `category/package_name` directory.

This is a scenario where there is a new commit in the same day or the SHA was incorrect.

```
2024-11-20 04:05:58.230 | DEBUG    | livecheck.main:get_props:101 - search_dir=/var/db/repos/tatsh
2024-11-20 04:05:58.260 | DEBUG    | livecheck.main:get_props:151 - Parsed path: /Sg4Dylan/vapoursynth-fsrcnn-ncnn-vulkan/archive/c9dad37dad64346afdff925fb409e6c6005b719a.tar.gz
2024-11-20 04:05:58.260 | DEBUG    | livecheck.main:main:423 - Fetching https://github.com/Sg4Dylan/vapoursynth-fsrcnn-ncnn-vulkan/commits/master.atom
2024-11-20 04:05:58.260 | DEBUG    | livecheck.main:main:431 - Adding Accept header for XML
2024-11-20 04:05:58.634 | DEBUG    | livecheck.main:do_main:292 - Using RE: "<id>tag:github.com,2008:Grit::Commit/([0-9a-f]{40})[0-9a-f]*</id>"
2024-11-20 04:05:58.634 | DEBUG    | livecheck.main:do_main:303 - Result count: 5
2024-11-20 04:05:58.634 | DEBUG    | livecheck.main:do_main:306 - re.findall() -> "0a721959116f6f86dbd13bda70a5fd6a883938f4"
2024-11-20 04:05:58.635 | DEBUG    | livecheck.main:do_main:323 - Attempted to fix top_hash date but it failed. Ignoring this error.
2024-11-20 04:05:58.635 | DEBUG    | livecheck.main:do_main:324 - top_hash = 0a721959116f6f86dbd13bda70a5fd6a883938f4
2024-11-20 04:05:58.635 | DEBUG    | livecheck.main:do_main:325 - Comparing current ebuild version c9dad37dad64346afdff925fb409e6c6005b719a with live version 0a721959116f6f86dbd13bda70a5fd6a883938f4
/var/db/repos/tatsh/media-video/vapoursynth-fsrcnn-ncnn-vulkan/vapoursynth-fsrcnn-ncnn-vulkan-20241112.ebuild -> vapoursynth-fsrcnn-ncnn-vulkan-20241112-r1.ebuild
```

The line showing the rename shows the bad target path.

**Comment by inode64** [comment]
Ok, nothing else I can do, I'll check it and solve it.

**Comment by inode64** [comment]
* Solve error when the hosts.yml file does not exist and without a manifest file for livechecks
* Add more types of variables to obtain the old sha 
* Remove variable prefixes in do_main, no longer used
* If the ebuild matches, use a new revision -r(+1)
* Unify the code to show the new ebuilds
* Moves the process of obtaining the regex within do_main, there may be code that does not use regex (jetbrains, pecl, sf, etc.) and thus eliminates the need to download the url
* Fix previous issues with packages
<pre>
2024-11-24 13:26:02.764 | INFO     | Processing dev-python/pymatting version 1.1.12_p20240409
dev-python/pymatting: 1.1.12_p20240409 (a073421a31afd0d53b7ff387ad2b361645b446be) -> 20241110 (afd2dec073cb08b8119300feec18c179a9d436f3)
2024-11-24 13:25:37.823 | INFO     | Processing media-video/vapoursynth-fsrcnn-ncnn-vulkan version 20200229
media-video/vapoursynth-fsrcnn-ncnn-vulkan: 20200229 (c9dad37dad64346afdff925fb409e6c6005b719a) -> 20241112 (0a721959116f6f86dbd13bda70a5fd6a883938f4)
2024-11-24 13:25:15.202 | INFO     | Processing dev-util/spirv-cross version 20240910
dev-util/spirv-cross: 20240910 (f84c1fbe3cbb69fa99e9115f658e52df0e95f1bf) -> 20241113 (9040e0d25dc545a6d725276bdbd0362791c81f14)
</pre>
I have checked with my repository and with the tatsh-overlay

**Comment by Tatsh** [comment]
The bug I mentioned is still there.

```
2024-11-24 07:51:56.061 | DEBUG    | livecheck.main:get_props:101 - search_dir=/var/db/repos/tatsh
DEBUG:asyncio:Using selector: EpollSelector
DEBUG:livecheck.utils.portage:Checking: /var/db/repos/tatsh == /var/db/repos/tatsh ?
2024-11-24 07:51:56.066 | DEBUG    | livecheck.main:get_props:151 - Parsed path: /Sg4Dylan/vapoursynth-fsrcnn-ncnn-vulkan/archive/c9dad37dad64346afdff925fb409e6c6005b719a.tar.gz
2024-11-24 07:51:56.066 | DEBUG    | livecheck.main:main:423 - Fetching https://github.com/Sg4Dylan/vapoursynth-fsrcnn-ncnn-vulkan/commits/master.atom
2024-11-24 07:51:56.066 | DEBUG    | livecheck.main:main:431 - Adding Accept header for XML
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): github.com:443
DEBUG:urllib3.connectionpool:https://github.com:443 "GET /Sg4Dylan/vapoursynth-fsrcnn-ncnn-vulkan/commits/master.atom HTTP/11" 200 4191
2024-11-24 07:51:56.278 | DEBUG    | livecheck.main:do_main:292 - Using RE: "<id>tag:github.com,2008:Grit::Commit/([0-9a-f]{40})[0-9a-f]*</id>"
2024-11-24 07:51:56.278 | DEBUG    | livecheck.main:do_main:303 - Result count: 5
2024-11-24 07:51:56.278 | DEBUG    | livecheck.main:do_main:306 - re.findall() -> "0a721959116f6f86dbd13bda70a5fd6a883938f4"
2024-11-24 07:51:56.279 | DEBUG    | livecheck.main:do_main:323 - Attempted to fix top_hash date but it failed. Ignoring this error.
2024-11-24 07:51:56.279 | DEBUG    | livecheck.main:do_main:324 - top_hash = 0a721959116f6f86dbd13bda70a5fd6a883938f4
2024-11-24 07:51:56.280 | DEBUG    | livecheck.main:do_main:325 - Comparing current ebuild version c9dad37dad64346afdff925fb409e6c6005b719a with live version 0a721959116f6f86dbd13bda70a5fd6a883938f4
/var/db/repos/tatsh/media-video/vapoursynth-fsrcnn-ncnn-vulkan/vapoursynth-fsrcnn-ncnn-vulkan-20241112.ebuild -> vapoursynth-fsrcnn-ncnn-vulkan-20241112-r1.ebuild
```

Please try to re-create this by creating the same ebuild as what is in my repository in your own repository or you can clone my repository (but it has to be in installed repositories).

https://github.com/Tatsh/tatsh-overlay/blob/master/media-video/vapoursynth-fsrcnn-ncnn-vulkan/vapoursynth-fsrcnn-ncnn-vulkan-20241112.ebuild

In your repo for testing:

```shell
mkdir -p media-video/vapoursynth-fsrcnn-ncnn-vulkan
wget 'https://raw.githubusercontent.com/Tatsh/tatsh-overlay/refs/heads/master/media-video/vapoursynth-fsrcnn-ncnn-vulkan/vapoursynth-fsrcnn-ncnn-vulkan-20241112.ebuild' -O media-video/vapoursynth-fsrcnn-ncnn-vulkan/vapoursynth-fsrcnn-ncnn-vulkan-20241112.ebuild'
ebuild media-video/vapoursynth-fsrcnn-ncnn-vulkan/vapoursynth-fsrcnn-ncnn-vulkan-20241112.ebuild manifest

# Run this branch of livecheck
livecheck -d -a vapoursynth-fsrcnn-ncnn-vulkan

# This should not output a file
ls *.ebuild
```

Please share the log.

**Comment by inode64** [comment]
<pre>
$ python -m livecheck -W /usr/portage/local/tatsh-overlay/media-video/vapoursynth-fsrcnn-ncnn-vulkan/ -p -d -a
!!! Found 2 make.conf files, using both '/etc/make.conf' and '/etc/portage/make.conf'
2024-11-24 14:06:40.536 | INFO     | livecheck.main:main:544 - search_dir=/usr/portage/local/tatsh-overlay/media-video/vapoursynth-fsrcnn-ncnn-vulkan repo_root=/usr/portage/local/tatsh-overlay repo_name=tatsh-overlay
DEBUG:asyncio:Using selector: EpollSelector
2024-11-24 14:06:40.539 | INFO     | livecheck.main:get_props:109 - Found 1 ebuilds
2024-11-24 14:06:40.539 | INFO     | livecheck.main:get_props:129 - Processing media-video/vapoursynth-fsrcnn-ncnn-vulkan version 20200229
2024-11-24 14:06:40.539 | DEBUG    | livecheck.main:get_props:165 - Parsed path: /Sg4Dylan/vapoursynth-fsrcnn-ncnn-vulkan/archive/c9dad37dad64346afdff925fb409e6c6005b719a.tar.gz
2024-11-24 14:06:40.540 | DEBUG    | livecheck.main:do_main:318 - Fetching https://github.com/Sg4Dylan/vapoursynth-fsrcnn-ncnn-vulkan/commits/master.atom
2024-11-24 14:06:40.540 | DEBUG    | livecheck.main:do_main:326 - Adding Accept header for XML
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): github.com:443
DEBUG:urllib3.connectionpool:https://github.com:443 "GET /Sg4Dylan/vapoursynth-fsrcnn-ncnn-vulkan/commits/master.atom HTTP/11" 200 4191
2024-11-24 14:06:40.867 | DEBUG    | livecheck.main:do_main:338 - Using RE: "<id>tag:github.com,2008:Grit::Commit/([0-9a-f]{40})[0-9a-f]*</id>"
2024-11-24 14:06:40.868 | DEBUG    | livecheck.main:do_main:350 - Result count: 5
2024-11-24 14:06:40.868 | DEBUG    | livecheck.main:do_main:354 - re.findall() -> "0a721959116f6f86dbd13bda70a5fd6a883938f4"
2024-11-24 14:06:40.868 | DEBUG    | livecheck.main:do_main:364 - Use updated date 20241112 for commit 0a721959116f6f86dbd13bda70a5fd6a883938f4
2024-11-24 14:06:40.869 | DEBUG    | livecheck.main:do_main:379 - Attempted to fix top_hash date but it failed. Ignoring this error.
2024-11-24 14:06:40.869 | DEBUG    | livecheck.main:do_main:380 - top_hash = 0a721959116f6f86dbd13bda70a5fd6a883938f4
2024-11-24 14:06:40.869 | DEBUG    | livecheck.main:do_main:388 - Comparing current ebuild version c9dad37dad64346afdff925fb409e6c6005b719a with live version 0a721959116f6f86dbd13bda70a5fd6a883938f4
2024-11-24 14:06:40.869 | DEBUG    | livecheck.main:do_main:393 - Updating ebuild /usr/portage/local/tatsh-overlay/media-video/vapoursynth-fsrcnn-ncnn-vulkan/vapoursynth-fsrcnn-ncnn-vulkan-20200229.ebuild to /usr/portage/local/tatsh-overlay/media-video/vapoursynth-fsrcnn-ncnn-vulkan/vapoursynth-fsrcnn-ncnn-vulkan-20241112.ebuild
media-video/vapoursynth-fsrcnn-ncnn-vulkan: 20200229 (c9dad37dad64346afdff925fb409e6c6005b719a) -> 20241112 (0a721959116f6f86dbd13bda70a5fd6a883938f4)
/usr/portage/local/tatsh-overlay/media-video/vapoursynth-fsrcnn-ncnn-vulkan/vapoursynth-fsrcnn-ncnn-vulkan-20200229.ebuild -> /usr/portage/local/tatsh-overlay/media-video/vapoursynth-fsrcnn-ncnn-vulkan/vapoursynth-fsrcnn-ncnn-vulkan-20241112.ebuild
>>> Downloading 'http://distfiles.gentoo.org/distfiles/04/vapoursynth-fsrcnn-ncnn-vulkan-20241112.tar.gz'
--2024-11-24 14:06:40--  http://distfiles.gentoo.org/distfiles/04/vapoursynth-fsrcnn-ncnn-vulkan-20241112.tar.gz
Resolving distfiles.gentoo.org (distfiles.gentoo.org)... 109.61.88.171, 109.61.88.175, 2a02:6ea0:c300::17, ...
Connecting to distfiles.gentoo.org (distfiles.gentoo.org)|109.61.88.171|:80... connected.
HTTP request sent, awaiting response... 404 Not Found
2024-11-24 14:06:41 ERROR 404: Not Found.

>>> Downloading 'https://github.com/Sg4Dylan/vapoursynth-fsrcnn-ncnn-vulkan/archive/0a721959116f6f86dbd13bda70a5fd6a883938f4.tar.gz'
--2024-11-24 14:06:41--  https://github.com/Sg4Dylan/vapoursynth-fsrcnn-ncnn-vulkan/archive/0a721959116f6f86dbd13bda70a5fd6a883938f4.tar.gz
Resolving github.com (github.com)... 140.82.121.3
Connecting to github.com (github.com)|140.82.121.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://codeload.github.com/Sg4Dylan/vapoursynth-fsrcnn-ncnn-vulkan/tar.gz/0a721959116f6f86dbd13bda70a5fd6a883938f4 [following]
--2024-11-24 14:06:42--  https://codeload.github.com/Sg4Dylan/vapoursynth-fsrcnn-ncnn-vulkan/tar.gz/0a721959116f6f86dbd13bda70a5fd6a883938f4
Resolving codeload.github.com (codeload.github.com)... 140.82.121.10
Connecting to codeload.github.com (codeload.github.com)|140.82.121.10|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/x-gzip]
Saving to: ‘/usr/portage/distfiles/vapoursynth-fsrcnn-ncnn-vulkan-20241112.tar.gz.__download__’

/usr/portage/distfiles/vapoursynth-fsrcnn-ncnn-vulkan-20241112.tar.gz.     [ <=>                                                                                                                                                                       ]  79,69K  --.-KB/s    in 0,08s   

2024-11-24 14:06:42 (1010 KB/s) - ‘/usr/portage/distfiles/vapoursynth-fsrcnn-ncnn-vulkan-20241112.tar.gz.__download__’ saved [81607]

>>> Creating Manifest for /usr/portage/local/tatsh-overlay/media-video/vapoursynth-fsrcnn-ncnn-vulkan
</pre>

<pre>
$ ls /usr/portage/local/tatsh-overlay/media-video/vapoursynth-fsrcnn-ncnn-vulkan/
Manifest  files  metadata.xml  vapoursynth-fsrcnn-ncnn-vulkan-20241112.ebuild

</pre>

**Comment by inode64** [comment]
[tatsh.log](https://github.com/user-attachments/files/17893508/tatsh.log)

Include log

**Comment by inode64** [comment]
I have seen a problem, then I will upload the correction, after making the adjustments and tests

**Comment by Tatsh** [comment]
Clean up your copy of my overlay and try this and send output:

```
python -m livecheck -W /usr/portage/local/tatsh-overlay/ -p -d -a vapoursynth-fsrcnn-ncnn-vulkan
ls /usr/portage/local/tatsh-overlay/*.ebuild
```

**Comment by Tatsh** [comment]
Ignore the issue with -r1 files being placed in the root of the repo. That issue exists even in the last release. https://github.com/Tatsh/livecheck/issues/244

**Comment by Tatsh** [comment]
Please either redo this PR or fix your branch so it does not need a merge. The project was renamed from `portage-livecheck` to `livecheck`.

**Comment by inode64** [comment]
now it seems to work
tonight I will redo the pull request

<pre>
2024-11-25 07:13:47.382 | INFO     | livecheck.main:get_props:109 - Found 1 ebuilds
2024-11-25 07:13:47.383 | INFO     | livecheck.main:get_props:129 - Processing media-video/vapoursynth-fsrcnn-ncnn-vulkan version 20241112
2024-11-25 07:13:47.383 | DEBUG    | livecheck.main:get_props:165 - Parsed path: /Sg4Dylan/vapoursynth-fsrcnn-ncnn-vulkan/archive/c9dad37dad64346afdff925fb409e6c6005b719a.tar.gz
2024-11-25 07:13:47.383 | DEBUG    | livecheck.main:do_main:324 - Fetching https://github.com/Sg4Dylan/vapoursynth-fsrcnn-ncnn-vulkan/commits/master.atom
2024-11-25 07:13:47.383 | DEBUG    | livecheck.main:do_main:332 - Adding Accept header for XML
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): github.com:443
DEBUG:urllib3.connectionpool:https://github.com:443 "GET /Sg4Dylan/vapoursynth-fsrcnn-ncnn-vulkan/commits/master.atom HTTP/11" 200 4191
2024-11-25 07:13:47.669 | DEBUG    | livecheck.main:do_main:344 - Using RE: "<id>tag:github.com,2008:Grit::Commit/([0-9a-f]{40})[0-9a-f]*</id>"
2024-11-25 07:13:47.669 | DEBUG    | livecheck.main:do_main:356 - Result count: 5
2024-11-25 07:13:47.669 | DEBUG    | livecheck.main:do_main:360 - re.findall() -> "0a721959116f6f86dbd13bda70a5fd6a883938f4"
2024-11-25 07:13:47.669 | DEBUG    | livecheck.main:do_main:370 - Use updated date 20241112 for commit 0a721959116f6f86dbd13bda70a5fd6a883938f4
2024-11-25 07:13:47.670 | DEBUG    | livecheck.main:do_main:385 - Attempted to fix top_hash date but it failed. Ignoring this error.
2024-11-25 07:13:47.670 | DEBUG    | livecheck.main:do_main:386 - top_hash = 0a721959116f6f86dbd13bda70a5fd6a883938f4
2024-11-25 07:13:47.670 | DEBUG    | livecheck.main:do_main:394 - Comparing current ebuild version c9dad37dad64346afdff925fb409e6c6005b719a with live version 0a721959116f6f86dbd13bda70a5fd6a883938f4
2024-11-25 07:13:47.670 | DEBUG    | livecheck.main:do_main:400 - Updating ebuild /usr/portage/local/tatsh-overlay/media-video/vapoursynth-fsrcnn-ncnn-vulkan/vapoursynth-fsrcnn-ncnn-vulkan-20241112.ebuild to /usr/portage/local/tatsh-overlay/media-video/vapoursynth-fsrcnn-ncnn-vulkan/vapoursynth-fsrcnn-ncnn-vulkan-20241112.ebuild
media-video/vapoursynth-fsrcnn-ncnn-vulkan: 20241112 (c9dad37dad64346afdff925fb409e6c6005b719a) -> 20241112-r1 (0a721959116f6f86dbd13bda70a5fd6a883938f4)
</pre>

**Comment by inode64** [comment]
Continue in #241

**Pull request closed without merging**